src/compiler/parser.ts

```diff
@@ -3210,6 +3285,155 @@ module ts {
             node.name = parseRightSideOfDot(/*allowIdentifierNames*/ true);
             return finishNode(node);
         }
+        
+        // Return a string representation of an EntityName
+        function entityNameToString(entityName: EntityName): string {
+            if (entityName.kind === SyntaxKind.Identifier) {
+                return (<Identifier>entityName).text;
+            } else {
+                let qualifiedName = <QualifiedName>entityName;
+                return entityNameToString(qualifiedName.left) + '.' + entityNameToString(qualifiedName.right);
+            }
+            return "";
+        }
+
+        function parseJsxChild(): JsxText | JsxExpression | JsxElement {
+            let result: JsxText | JsxExpression | JsxElement = undefined;
+            switch (token) {
+                case SyntaxKind.JsxText:
+                    result = <JsxText>createNode(SyntaxKind.JsxText);
+                    token = scanner.scanJsxToken();
+                    result = finishNode(result);
+                    break;
+                case SyntaxKind.OpenBraceToken:
+                    result = parseJsxExpression();
+                    break;
+                default:
+                    Debug.assert(token === SyntaxKind.LessThanToken);
+                    result = parseJsxElement();
+                    break;
+            }
+            token = scanner.reScanJsxToken();
+            Debug.assert(result !== undefined, "parsed some JSX child");
+            return result;
+        }
+
+        function parseJsxElement(): JsxElement {
+            let node = <JsxElement>createNode(SyntaxKind.JsxElement);
+            node.openingElement = parseJsxOpeningElement();
+
+            let tagName = entityNameToString(node.openingElement.tagName);
+
+            if (node.openingElement.isSelfClosing) {
+                node.children = <NodeArray<JsxText | JsxExpression | JsxElement>>[];

```

Comment:
If you go this route, this node still needs to be correctly set up.  i.e. it's pos/end must be correct.  In general this is why manually creating arrays in the parser is not a good idea.